### ISSUE TITLE
Use Pageant named pipe as default and fallback to shared memory if unavailable

### ISSUE THREAD

**Issue by pzystorm** [keepassxreboot/keepassxc#9107]
I have a Laptop with Windows 10 (22H2) and the newest versions of Pageant and KeePassXC, KeePassXC has been installed via Microsoft Store. And regarding Pageant I have tried two versions: The exe file directly from the maintainer's website. And I have installed the official WinSCP install package which contains a pageant.

But it does not matter which pageant I start the KeePassXC is not able to communicate with the Pageant.
It always states:
```
Agent refused this identity. Possible reasons include:
The key has already been added.
A confirmation request is not supported by the agent (check options).
```

But afterwards the key is not in the pageant.
I think the company policy prevents this connection as I am connected to a MS company domain server.

I have a second laptop, a private one, also with Win 10 22H2, and this is working very well with Pageant (downloaded exe file from the maintainer's website) and the KeePassXC from the MS store.
So I think the issue must the at the company policy side.

Can someone tell me what our Domain Admins need to enable / allow / install that KeePassXC can communicate with Pageant?
Is this some DLL call? Named pipes? IPC?

**Comment by droidmonkey** [comment]
You need to allow access to this pipe:  `\\\\.\\pipe\\openssh-ssh-agent`

**Comment by pzystorm** [comment]
Thanks a lot for your quick answer. But I think this is the answer in the case I am using the openssh agent in Windows, correct?
I have looked in the source code and found this: it is just using SendMessage sending the message to the Pageant window:
https://github.com/keepassxreboot/keepassxc/blob/develop/src/sshagent/SSHAgent.cpp#L242

But anyway:
Since some years pageant also supports named pipes. In the Powershell I get:

PS C:\Users\myusername> [System.IO.Directory]::GetFiles("\\.\\pipe\\") | select-string "page"
\\.\\pipe\\pageant.myforename.mylastname.a1fdaff21c1e....blehblah...4c4071bd80c29

Would it be possible to support this instead of the SendMessage thingy to talk to Pageant?

Here is the official doc for this:
https://www.chiark.greenend.org.uk/~sgtatham/putty/wishlist/pageant-named-pipe.html

**Comment by droidmonkey** [comment]
Yes it is true that Paegent uses SendMessage to share data. I am not aware of group policies that restrict that, however. Seems like a bad idea that will break applications. Your corporation may need to place KeePassXC into a trusted state, perhaps AppLocker is responsible for the restrictions? 

@hifi what are your thoughts on using Pageants named pipe?

**Comment by hifi** [comment]
That makes sense because we already use it for OpenSSH for Windows, it's probably fairly trivial to add support for the named pipe. We can probably also make it auto-detect if it exists and use it if it does and have the shared memory method as a fallback so we don't need it to be configurable?

**Comment by droidmonkey** [comment]
I am down with that, reopening for that feature
